design-documents/asynchronous-operations/consumer-processes-improvements.md

```diff
@@ -0,0 +1,86 @@
+
+There is also a poison pill feature in Magento, which is basically just a random hash stored in the database in the table `queue_poison_pill`. If some configuration value gets changed using the Magento backend, the poison pill version is changed. 
+The running consumer processes check on this version, but only when a new message is in the queueu. If a consumer discovers a new message in the queue, it first checks if the poison pill version has changed. If the version wasn't changed, the message will be handled. If the version did get changed, the consumer will kill itself and the cronjob will spawn a new one on its next run and only then will handle that message.
+
+Magento 2.3.2 was the first Open Source version coming with some default features making actual use of the message queue system. These are tasks which are triggered using the backend of Magento which can potentially take a while to execute. In order to prevent the webserver or php-fpm to run against a timeout, it was chosen to send these tasks to the message queue system and let these tasks get executed asynchronously. Some examples are:
+
+- Generating coupon codes
+- Mass editing products
+- Exporting data
+- ...
+
+Currently in Magento 2.3.2, 4 consumer processes get spawned by the cron system. Each of these processes take memory and cpu and regularly queries the database (or RabbitMQ if that broker is being used) to see if new messages are available and then process them.
+
+
+## Problem 1: too little control
+
+### The problem
+
+There is currently too little control over these consumer processes.  
+What if people only very irregularly use one of these features mentioned above, let's say they only once a year export all their products to check their inventory levels.  
+Then you have some consumer process sitting there, doing nothing for 364 days in the year, wasting precious cpu cycles and taking up precious memory, until finally once a year the shopowner decides to execute a certain task it can execute.
+
+There is also the potential problem that the consumer process will take up more and more memory due to some memory leak in the code it executes.
+
+### The suggested solution
+
+The suggestion is to give more control per consumer to the shopowner or developers managing the shop.  
+We could add some additional options to the consumer processes to keep them more under control. Currently there is one available limit available: `max-messages`. If that number of messages was processed, the consumer will kill itself.  
+I'd like to suggest some other limits which we can set:
+
+- `max-idle-time`: if no message was being handled in xx seconds, then kill yourself
+- `max-time`: after xx total seconds, and after you are done with handling the current message, kill yourself
+- `memory-limit`: if xx MB's of memory is being taken up by the process and after you are done with handling the current message, kill yourself (a way to work around potential memory leaks)
```

Comment:
Thats should not be a case, cause if there a memory leak then its an architectural/core bug that have to be fixed on code level and not every time kill the process. Its not solving the problem, just reseting it.

Reply:
True, it's not strictly needed, got the idea from https://symfony.com/doc/current/messenger.html#deploying-to-production. The idea is that these things are configurable and everybody can choose to use it or not. Shall I leave it for now?

Reply:
Lets have a look at another opinions :-)

Reply:
Thanks @tariqjawed83 for the feedback!

Right, ok, I'll remove `memory-limit` and `max-time` again from the proposal.
I was just looking at what `symfony/messenger` was doing and and took over some features from them.

Reply:
Removed these two options.